events via MLS proposals (#5)
bifurcation/ietf-mimi-protocol: draft-ralston-mimi-protocol.md

A room's state is modified, used, and retrieved through *events*. Some events
are fanned out to other participating servers while other events are operations
performed exclusively with the hub server for a room.

Events that change the state of the room are implemented through MLS proposals
as defined in {{!RFC9420}}, thus allowing the underlying MIMI DS protocol to
anchor the current room state cryptographically. MLS proposals are signed,
allowing every recipient in their path to verify their authenticity.

This document defines additional events to encapsulate MIMI DS protocol

> messages, for example, to add clients to a room's underlying MLS group or to

Reviewer: ```suggestion
messages, for example, to
```

   // Additional fields may be present as dependent on event type.
   select (Event.type) {
      case "m.room.user":
         // MLSMessage containing a UserEvent proposal

>          MLSMessage user_event_proposal; // see later in doc

Reviewer: ```suggestion
         Proposal user_event_proposal; // see later in doc
```

The "origin server" of an event is the server implied by the `sender` field.

## Room state

The state of a room consists of the room's RoomID, its policy, and the
participant list (including the role and participation state of each
participant). Also associated with the room is the MLS group managed by the MIMI
DS protocol, which anchors the room state cryptographically as part of the group
state.

While (through the MIMI DS protocol) all parties involved in a room agree on the
room's state, the Hub is the arbiter that decides if a state change is valid.
All state-changing events are sent to the Hub, checked for their validity and
policy conformance before they are forwarded to any follower servers.

As soon as the Hub accepts an event that changes the room state, its effect is
applied to the room state and future events are validated in the context of that
new state.

The room state is thus changed based on events, even if the MLS proposal
implementing the event was not yet committed by a client. Note that this only
applies to events changing the room state, but not for MIMI DS specific events
that change the group state. For more information on the proposal-commit

> paradigm and the role of the MIMI DS protocol see {{mimi-ds}}.

Reviewer: This is still very awkward. I'd just add a TODO to reword it and I can work on some proposed text later.

## Cryptographic room representation {#mimi-ds}

Each room is represented cryptographically by an MLS group and the Hub that

> manages the room acts as the delivery service (DS) as defined in the MIMI DS

Reviewer: Isn't the DS supposed to be spread across all the clients and all the providers?  Maybe something like the "the arbiter of the delivery service (DS)"?

Feel free to ignore this comment until later.

Author: Fair point. I'll clarify that in the next PR.

struct {
  opaque room_id;
  ParticipantData participants<V>;
  // TODO: Add any remaining room data
} RoomState;
~~~

As part of the MIMI DS protocol, clients create commits to update the group
state, which are then included in MIMI DS specific events. The time between two
commits denotes an epoch.

Whenever a client creates a commit, it MUST include all valid proposals accepted
by the Hub during the current epoch. This includes both proposals that carry
room-state changes, as well as proposals sent as part of MIMI DS events.

Note that the validity of a proposal depend on the current room state, which may

> change during anepoch based on room-state changing events. The changes of these

Reviewer 2: ```suggestion
change during an epoch based on room-state changing events. The changes of these
```

events are applied to the room state even if the commits that carry the event
information have not yet been committed.

### Authenticating proposals

The MLS specification {{!RFC9420}} requires that MLS proposals from the Hub and
from follower servers (external senders in MLS terminology) be authenticated
using key material contained in the `external_senders` extension of the MLS
group. Each MLS group associated with a MIMI room MUST therefore contain an
`external_senders` extension. That extension MUST contain at least the
Certificate of the Hub.

When a user from a follower server becomes a participant in the room, the
Certificate of the follower server MAY be added to the extension. When the last
participant belonging to a follower server leaves the room, the certificate of
that user MUST be removed from the list. Changes to the `external_senders`
extension only take effect when the MLS proposal containing the event is

> committed by a MIMI DS commit. See {{ev-mroomuser}} for more information.

Reviewer 2: Is this to provide the follower an ability to propose removal of their own users' clients? It feels like a path towards hub portability, which I'm very happy to include, though not sure if we want to open that can of worms yet.

Author: Kind of. It's for the follower server to be able to send signaling messages at all (or rather for everyone else to be able to authenticate them). This is a limitation stemming from the way that proposals from external senders (i.e. senders who are not group members) are authenticated. My thinking was that if a user is part of a room, its server might want to be able to send signaling messages independent of the user's clients. If it wants to do that, it needs to be added to the list. This has to happen before the server can send any signaling messages (or at least those implemented by MLS proposals) and it can only happen via commit by a client.

**Event type**: `m.room.user`

> An `m.room.user` event can be used to change the participation state of a room's
> participant.

Reviewer 2: ```suggestion
An `m.room.user` event can be used to change the participation state of a user.
```

to avoid "participation" appearing twice.

> > **TODO**: Do we also want this to be able to change a participant's role?

Reviewer 2: Role and participation are very different concepts imo. They could be part of the atomic operations though, where we add a user, their clients, and promote them to admin in a single transaction/frame. 

Definitely worth leaving as a TODO for now 😇

It is transported via an MLS proposal of type UserEvent. If the event adds a
user to the room and it is the first user in the room that belongs to the
sending follower server, the UserEvent MAY contain the Certificate that can be
used to validate external proposals from that follower server. If it does, the
commit that contains the proposal adds the Certificate to `external_senders`
extension of the underlying MLS group.

If the event removes the last user of a follower server from a room, the commit
that contains the MLS proposal that carries the event removes the Certificate of
that follower server from the extension.

> **TODO**: This proposal needs to be added to the IANA proposal list, or
> specified as an extension proposal as specified in the MLS extensions
> document. We might want to have one MIMIProposal type that in turn can
> encapsulate more than just this event.

~~~
struct {
   // The user ID being affected by this participation state change.
   opaque targetUserId;

   // The new participation state for the target user.

>    Participan state;

Reviewer 2: accidental edit?
```suggestion
   ParticipationState state;
```
